### ISSUE TITLE
torch.norm's inf variant is incorrect when given non-contiguous inputs

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#67517]
Per title

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @jianyuh @mruberry @walterddr @IvanYashchuk @xwang233

**Comment by krshrimali** [comment]
While investigating this with an independent script:

<details>

```python
import torch
from torch.testing._internal.common_utils import noncontiguous_like
from torch.testing._internal.common_methods_invocations import SampleInput
from torch.testing import make_tensor

# Setting the seed gives True, otherwise False (output of the last statement `torch.allclose`)
# torch.manual_seed(42)
x = make_tensor((5, 5), dtype=torch.complex64, device='cpu', requires_grad=True)
s = SampleInput(x, args=(float('inf'),))
t_inp, t_args = s.input, s.args
n_inp, n_args, n_kwargs = s.noncontiguous()

expected = torch.norm(t_inp, *t_args)
actual = torch.norm(n_inp, *n_args)

expected_backward_tensor = expected
actual_backward_tensor = actual
grad_for_expected = torch.randn_like(expected_backward_tensor)
grad_for_actual = noncontiguous_like(grad_for_expected)

expected_backward_tensor.backward(grad_for_expected)
actual_backward_tensor.backward(grad_for_actual)

expected_grad = t_inp.grad
actual_grad = n_inp.grad
print(torch.allclose(expected_grad, actual_grad))
```

</details>

When the random seed is set, `expected_grad` and `actual_grad` are equal (`torch.allclose` returns `True`), otherwise they aren't. Though, setting a wrapper to set random seed like it's done for `dropout` (https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L11376) doesn't work in this case. The test still fails...

Diving deeper into this, https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L198 - the output for contiguous and non-contiguous inputs is different here for `self.abs() == norm`.

For contiguous inputs:

```python
self.abs():
  9.5151   1.9681   4.3448   5.6507  10.1162
  4.6900   8.6012   5.2387   4.4827   3.0662
  5.9427   6.2884   5.1211  10.0379   8.5886
  3.6435   9.7829   2.9180   5.3512   5.5322
  4.1583   3.8123   8.9807   7.2914  10.0650
[ CPUFloatType{5,5} ]

norm: 10.1162
[ CPUFloatType{1,1} ]

self.abs().eq(norm):

 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
```

For non-contiguous inputs:

```python
self.abs(): 
  9.5151   1.9681   4.3448   5.6507  10.1162
  4.6900   8.6012   5.2387   4.4827   3.0662
  5.9427   6.2884   5.1211  10.0379   8.5886
  3.6435   9.7829   2.9180   5.3512   5.5322
  4.1583   3.8123   8.9807   7.2914  10.0650
[ CPUFloatType{5,5} ]

norm: 10.1162
[ CPUFloatType{1,1} ]

self.abs().eq(norm):
 0  0  0  0  1
 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
 0  0  0  0  0
```

Please note the difference in the output for `self.abs().eq(norm)` for both the cases. Not sure if this helps, but thought I'll share what I've noted (so far), here.

**Comment by ngimel** [comment]
I think we already had this problem, it's related to using vectorized/non-vectorized kernels that compute `abs` slightly differently.

**Comment by kurtamohler** [comment]
Yes, this sounds like it's related to vectorization.

You can change the vectorization that PyTorch uses by setting the `ATEN_CPU_CAPABILITY` environment variable before running a script. These values can be used:
* `ATEN_CPU_CAPABILITY=default` - turns off AVX/AVX2
* `ATEN_CPU_CAPABILITY=avx` - turns on AVX
* `ATEN_CPU_CAPABILITY=avx2` - turns on AVX2 (I think this one is usually on by default)

**Comment by kurtamohler** [comment]
Running the script that @krshrimali  posted above, the contiguous and non-contiguous results match each other when `ATEN_CPU_CAPABILITY=default` is set
